src/cargo/ops/cargo_rustc/mod.rs

```diff
@@ -63,89 +65,97 @@ pub fn compile_targets<'a, 'cfg: 'a>(targets: &[(&'a Target, &'a Profile)],
                                      build_config: BuildConfig,
                                      profiles: &'a Profiles)
                                      -> CargoResult<Compilation<'cfg>> {
-    if targets.is_empty() {
-        return Ok(Compilation::new(pkg, config))
-    }
 
-    debug!("compile_targets: {}", pkg);
+    debug!("compile_targets: {}", pkg_targets.iter().map(|&(ref p, _)| p.name())
+                                                    .collect::<Vec<_>>().connect(", "));
 
     try!(links::validate(deps));
 
     let dest = if build_config.release {"release"} else {"debug"};
-    let root = if resolve.root() == pkg.package_id() {
-        pkg
-    } else {
-        deps.iter().find(|p| p.package_id() == resolve.root()).unwrap()
-    };
+    let root = deps.iter().find(|p| p.package_id() == resolve.root()).unwrap();
     let host_layout = Layout::new(config, root, None, &dest);
     let target_layout = build_config.requested_target.as_ref().map(|target| {
         layout::Layout::new(config, root, Some(&target), &dest)
     });
 
     let mut cx = try!(Context::new(resolve, sources, deps, config,
-                                   host_layout, target_layout, pkg,
+                                   host_layout, target_layout,
                                    build_config, profiles));
 
     let mut queue = JobQueue::new(cx.resolve, deps, cx.jobs());
 
-    // Prep the context's build requirements and see the job graph for all
-    // packages initially.
     {
         let _p = profile::start("preparing build directories");
-        try!(cx.prepare(pkg, targets));
-        prepare_init(&mut cx, pkg, &mut queue, &mut HashSet::new());
-        custom_build::build_map(&mut cx, pkg, targets);
+        // Prep the context's build requirements and see the job graph for all
+        // packages initially.
+        for &(pkg, ref targets) in pkg_targets {
+            try!(cx.prepare(pkg, targets));
+            prepare_init(&mut cx, pkg, &mut queue, &mut HashSet::new());
+            custom_build::build_map(&mut cx, pkg, targets);
```

Comment:
This function isn't currently really designed to be called more than once, and actually taking a look at these... most of them shouldn't actually work after being called more than once. I think the passing of multiple targets will need to be threaded a bit deeper to each of these so the loop can happen farther down rather than up here.

It'd be fine to define a type alias for this package of "things to build" if typing out the slice and types gets too verbose.